async_upnp_client/ssdp_listener.py

```diff
@@ -100,30 +100,20 @@ def __repr__(self) -> str:
         return f"<{type(self).__name__}({self.udn})>"
 
 
-def headers_differ(current_headers: SsdpHeaders, new_headers: SsdpHeaders) -> bool:
-    """Compare headers to see if anything interesting has changed."""
-    current_filtered = {
-        key: value
-        for key, value in current_headers.items()
-        if not key.startswith("_") and key.upper() not in IGNORED_HEADERS
-    }
-    new_filtered = {
-        key: value
-        for key, value in new_headers.items()
-        if not key.startswith("_") and key.upper() not in IGNORED_HEADERS
-    }
-    if _LOGGER.level <= logging.DEBUG:
-        diff_values = {
-            k: (
-                current_filtered.get(k),
-                new_filtered.get(k),
+def same_headers_differ(current_headers: SsdpHeaders, new_headers: SsdpHeaders) -> bool:
+    """Compare headers present in both to see if anything interesting has changed."""
+    for header, new_value in new_headers.items():
+        if header.startswith("_") or header.upper() in IGNORED_HEADERS:
```

Comment:
To get the best performance should convert every key and always use that in `CaseInsensitiveDict`. I.e., discard the original key and keep the new one. But that would mean loosing the original data from the responses.

Wrt performance this is the way to go since this will eliminate conversions at a later stage.
But I get the feeling that this might shoot us in the foot somewhere later. Just a feeling, nothing concrete.

Reply:
https://openconnectivity.org/upnp-specs/UPnP-arch-DeviceArchitecture-v2.0-20200417.pdf refers to [RFC  2616](https://datatracker.ietf.org/doc/html/rfc2616.html) on page 21. [RFC 2616 Section 4.2](https://datatracker.ietf.org/doc/html/rfc2616.html#section-4.2) states (emphasis mine):
> HTTP header fields, which include general-header (section 4.5),
   request-header (section 5.3), response-header (section 6.2), and
   entity-header (section 7.1) fields, follow the same generic format as
   that given in Section 3.1 of RFC 822 [9]. Each header field consists
   of a name followed by a colon (":") and the field value. **Field names
   are case-insensitive.**

Are you ok with converting every key to uppercase @bdraco? I think that will save a transformation in `CaseInsensitiveDict.__eq__` and `CaseInsensitiveDict.__hash__`, next to to the `same_headers_differ` function.

Reply:
The choice upper or lower is arbitrary of course.